components/places/src/storage/history.rs

```diff
@@ -211,24 +211,108 @@ pub fn href_to_guid(db: &PlacesDb, url: &str) -> Result<Option<SyncGuid>> {
 
 /// Internal function for deleting a place, creating a tombstone if necessary.
 /// Assumes a transaction is already set up by the caller.
-fn do_delete_place_by_guid(db: &PlacesDb, guid: &SyncGuid) -> Result<()> {
+fn delete_all_visits_for_guid_in_tx(db: &PlacesDb, guid: &SyncGuid) -> Result<()> {
     // We only create tombstones for history which exists and with sync_status
     // == SyncStatus::Normal
-    let sql = "INSERT OR IGNORE INTO moz_places_tombstones (guid)
-               SELECT guid FROM moz_places
-               WHERE guid = :guid AND sync_status = :status";
-    db.execute_named_cached(sql, &[(":guid", guid), (":status", &SyncStatus::Normal)])?;
-    // and try the delete - it might not exist, but that's ok.
-    let delete_sql = "DELETE FROM moz_places WHERE guid = :guid";
-    db.execute_named_cached(delete_sql, &[(":guid", guid)])?;
+    let to_clean = db.conn().try_query_row(
```

Comment:
Another possible approach is to embrace the constraint violation :)  According to the sqlite docs, such a violation doesn't roll anything back, so we could, say, unconditionally and explicitly delete all visits, then (try to) delete the page, etc. The downsides of that approach are that it makes some triggers somewhat redundant and it might make error handling more complex (ie, we'd need to be extra careful about only handling the errors we expect). However, the upsides of that approach would be that if things get more complex in the future, we don't end up with more match arms here, each dealing with a somewhat complex subset of things to delete and the possibility of getting them wrong.

That said though, I'm not actually *suggesting* we do that, just throwing it out there to see what you think. I don't think this has crossed some complexity threshold yet, I'm just vaguely worried it might in the future. Happy to deal with that in that hypothetical future though :)

Reply:
That's a really good point! I thought about it more, and definitely agree that the `foreign_count` handling complicates things. I'm worried that treating a constraint violation as "delete the visits instead" might complicate things in the other direction, though—there are other reasons SQLite can return the error, and I'm a bit wary of relying on it, like you pointed out. We'd also still need to make sure to only insert tombstones for synced visits and URLs, which I think would still require another read.

So I decided to keep things as they are, but I'm happy to change them if you feel like it would be better! 🚀